homeassistant/components/openrgb/coordinator.py

```diff
@@ -0,0 +1,178 @@
+"""DataUpdateCoordinator for OpenRGB."""
+
+from __future__ import annotations
+
+import asyncio
+from datetime import timedelta
+import logging
+
+from openrgb import OpenRGBClient
+from openrgb.orgb import Device
+from openrgb.utils import RGBColor
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PORT, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.debounce import Debouncer
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import CONNECTION_ERRORS, DEFAULT_CLIENT_NAME, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+type OpenRGBConfigEntry = ConfigEntry[OpenRGBCoordinator]
+
+
+class OpenRGBCoordinator(DataUpdateCoordinator[dict[str, Device]]):
+    """Class to manage fetching OpenRGB data."""
+
+    client: OpenRGBClient
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=15),
+            config_entry=config_entry,
+            request_refresh_debouncer=Debouncer(
+                hass, _LOGGER, cooldown=0.5, immediate=False
+            ),
+        )
+        self.host = config_entry.data[CONF_HOST]
+        self.port = config_entry.data[CONF_PORT]
+        self.entry_id = config_entry.entry_id
+        self.server_address = f"{self.host}:{self.port}"
+        self._client_lock = asyncio.Lock()
+
+        # Register listener for Home Assistant stop event
+        self._stop_listener = hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, self.async_client_disconnect
+        )
```

Comment:
I don't believe we ever clean this up when the entry unloads, so you should wrap this in entry.async_on_unload instead